src/jdk.jpackage/macosx/classes/jdk/jpackage/internal/MacDmgPackageBuilder.java

```diff
@@ -54,18 +55,58 @@ List<Path> validatedDmgContent() {
 
     MacDmgPackage create() throws ConfigException {
         final var superPkgBuilder = pkgBuilder.pkgBuilder();
+        var installDirDeleteRoot = new Object() { Path path; };
         superPkgBuilder.installDir().ifPresent(installDir -> {
-            final var defaultInstallDirLocation = superPkgBuilder.defaultInstallDir().map(Path::getParent).orElseThrow();
-            if (!defaultInstallDirLocation.equals(installDir)) {
-                Log.info(I18N.format("message.install-dir-ignored", defaultInstallDirLocation));
-                superPkgBuilder.installDir(defaultInstallDirLocation);
+            if (Files.exists(installDir) && !Files.isDirectory(installDir)) {
+                    throw new RuntimeException(
+                             I18N.format("message.install-dir-invalid",
+                             installDir.toString()));
+            }
+
+            if (!Files.exists(installDir)) {
+                try {
+                    // Install dir can have multiple non-existing directories.
+                    // We need to find first non-existing directory, so we can
+                    // delete it.
+                    Path parent = installDir.getParent();
+                    Path deleteRoot = null;
+                    while (parent != null) {
+                        if (!Files.exists(parent.getParent())) {
+                            parent = parent.getParent();
+                        } else {
+                            deleteRoot = parent;
+                            break;
+                        }
+                    }
+
+                    if (parent == null) {
+                        throw new RuntimeException(
+                             I18N.format("message.install-dir-invalid",
+                             installDir.toString()));
+                    }
+
+                    installDirDeleteRoot.path = deleteRoot;
+
+                    try {
+                        Files.createDirectories(installDir);
+                    } catch (Exception ex) {
+                        throw new RuntimeException(
+                             I18N.format("message.install-dir-create",
+                             installDir.toString(),
+                             ex.getLocalizedMessage()));
+                    }
+                } catch (Exception ex) {
+                    Log.verbose(ex);
+                    throw new RuntimeException(ex);
+                }
```

Comment:
This code should be in a separate function covered with unit tests.

Reply:
Another issue with this code is that it creates a directory during the configuration phase. Files/directories should be created at the packaging phase. Another reason to move it into MacDmgPackager